table/block.cc

```diff
@@ -505,9 +513,50 @@ bool IndexBlockIter::PrefixSeek(const Slice& target, uint32_t* index) {
   }
 }
 
+// NOTE: in suffix seek, if the key is not found in the restart intervals,
+// the iterator will simply be set as "invalid", rather than returning
+// the key that is just pass the target key.
+// return value: found
+bool DataBlockIter::HashSeek(const Slice& target) {
+  assert(data_block_hash_index_);
+  Slice user_key = ExtractUserKey(target);
+  std::unique_ptr<DataBlockHashIndexIterator> data_block_hash_iter(
+      data_block_hash_index_->NewIterator(user_key));
+
+  for (; data_block_hash_iter->Valid(); data_block_hash_iter->Next()) {
+    uint32_t restart_index = data_block_hash_iter->Value();
+    SeekToRestartPoint(restart_index);
+    while (true) {
+      if (!ParseNextDataKey() || Compare(key_, user_key) >= 0) {
+        // key_ is internal key. If key_ is parsed successfully, it is either:
+        // 1) user_key + seq + type
+        // 2) (a key larger than user_key) + seq + type
+        break;
+      }
+    }
```

Comment:
Can you explain how it works?
Assume we have following keys, 1, 2, 3, ...., 99.
Let's say one bucket contains two keys, 22 and 88, and you are seeking to "22" and get into this bucket. Are you go through this for-loop from 22, 23 until 88?

Reply:
Short answer: the for loop will return if it can find the first `key_` that matches the `seek_key` without further looping on the rest of the keys in the bucket.

More details: the entries stored in the bucket are a checksum of the key and the residing restart interval index (or restart offset). 

Say `key01`, `key02`, ..., `key99` are in the block. `key22` and `key88` are hashed to the same bucket, among other keys. Assume they belong to different restart interval `R2` and `R8`. 

When seeking `key22`, we hash the key to get the bucket.  There can be many entries in the bucket. The iterator only emits entries that have a matching check sum, meaning a potential match. But we still have to linear check the restart interval to see if the key is really there or not, as some other key hashed to this bucket may happen to have an identical checksum.

In the example, say `key22` and `key88` have the same checksum. The outer `for-loop` is supposed to loop on the two entries `key22` and `key88`. Then in the inter `while-loop`, it find the entry for `key22`, then jump to the restart interval `R2` and performs linear search in the restart interval. I will then break out of the inner `while-loop` when `key_ >= user_key`. A following check make sure the user key matches. If it is a match, function returns without going for another for-loop on the bucket entry correspoding to `key88`.